src/Microsoft.DotNet.Wpf/src/Themes/Shared/Microsoft/Windows/Themes/DataGridHeaderBorder.cs

```diff
@@ -368,7 +368,7 @@ private static void CacheFreezable(Freezable freezable, int index)
 
             lock (_cacheAccess)
             {
-                if (_freezableCache[index] != null)
+                if (_freezableCache[index] == null)
```

Comment:
On going through the code I agree that the logic earlier was not right. But with the current changes, any call to CacheFreezable won't be able to over-write the cache value once stored. 

Probably need to look a bit more to check if that happens / is a valid scenario.

Reply:
I went through the code and this change is fine imo. As the CacheFreezable is only called when we get null from calling GetCachedFreezable. So, once cached we won't be overwriting the cache.